### ISSUE TITLE
Webpack error with apollo-engine-binary 

### ISSUE THREAD

**Issue by Steinarthor** [apollographql/apollo-engine-js#122]
Greetings!

I'm trying to implement Apollo Engine into my project but I'm having trouble building it. This is the message I keep getting when I try to build with Webpack. The project is in the early stages (I'm still learning).

Everything works just fine in dev mode when I use nodemon and babel, it's only when I try to build with Webpack when things go south for me.

What kind of loader is webpack asking about here ? 

**Error Message**

ERROR in ./node_modules/apollo-engine-binary-linux/engineproxy_linux_amd64
Module parse failed: Unexpected character '' (1:0)
You may need an appropriate loader to handle this file type.
(Source code omitted for this binary file)
 @ ./node_modules/apollo-engine/lib/index.js 57:30-99
 @ ./server.js
 @ multi babel-regenerator-runtime ./server.js

ERROR in ./node_modules/apollo-engine-binary-darwin/engineproxy_darwin_amd64
Module parse failed: Unexpected character '�' (1:0)
You may need an appropriate loader to handle this file type.
(Source code omitted for this binary file)
 @ ./node_modules/apollo-engine/lib/index.js 53:30-101
 @ ./server.js
 @ multi babel-regenerator-runtime ./server.js

ERROR in ./node_modules/apollo-engine-binary-windows/engineproxy_windows_amd64.exe
Module parse failed: Unexpected character '�' (1:2)
You may need an appropriate loader to handle this file type.
(Source code omitted for this binary file)
 @ ./node_modules/apollo-engine/lib/index.js 61:30-107
 @ ./server.js
 @ multi babel-regenerator-runtime ./server.js

**Webpack Config**

```
const path = require("path");

module.exports = {
  entry: ["babel-regenerator-runtime", "./server.js"],
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "./server.js"
  },
  target: "node",
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: ["@babel/preset-env"]
          }
        }
      }
    ]
  }
};
```
**Package Json**

```
 "scripts": {
    "dev": "nodemon ./server.js --exec babel-node",
    "build": "webpack",
    "test": "jest",
    "start": "node ./dist/server.js"
  }
```

Any help would be greatly appreciated.

**Comment by mikestaub** [comment]
I just hit this issue as well, and it blocked me from using apollo-engine. 

I think the solution would be to add the regex /apollo-engine-binary/ to your webpack config.externals, and use webpack-plugin-copy to move the actual binaries into your webpack bundle folder.

I tried this but still hit a runtime error. The package is using ```require.resolve()``` to find the binary paths, and webpack transforms this to a module number (instead of the path string), which causes the spawn process to fail.

If a different method was used to resolve the binary paths instead of require.resolve(), I think this would work.

**Comment by Steinarthor** [comment]
ahh! that explains it. I was going insane trying to figure this thing out. I really want to use apollo engine, but I'm afraid that I will also have to stop using it :(

**Comment by zionts** [comment]
Hey @Steinarthor @mikestaub thanks so much for bringing this to our attention! After looking into it some, I think this may be resolvable via the `loaders` field in the webpack configuration. We're looking into this more and will have a solution out as soon as we can :)

**Comment by zionts** [comment]
@Steinarthor @mikestaub We've got a fix for you!

Webpack (and rollup) will try to look for any imports (via import or require) and bundle them together into a single file. They assume they can bundle anything they reach and the problem you both ran into was webpack trying to inline the engineproxy binary that this package loads. To allow wepback to keep going and keep the the requires in place, you can specify an `externals` section to tell webpack to essentially ignore the binary as it's bundling.

[Externals](https://webpack.js.org/configuration/externals/) can take a number of different things, but we want to tell webpack to ignore apollo-engine-binary-* so we write a simple check as shown in the example below to say "please ignore this binary". You can also do something similar in rollup via rollup-node-resolve (which also has an externals section)

tldr; add this to your webpack config:
```js
externals: [
    function(context, request, callback) {
      if (request.indexOf("apollo-engine-binary") > -1) {
        return callback(null, "commonjs " + request);
      }
      callback();
    },
  ],
```

Thanks to @jbaxleyiii for his help on this!

**Comment by Steinarthor** [comment]
@zionts  - Thanks very much for your help mate <3

**Comment by mikestaub** [comment]
Thanks for looking into this @zionts , but even with your fix, what I said earlier still blocked me (about require.resolve). I decided to just include a regex to ignore all apollo-engine modules from my bundle, and then manually copy them over (with exec perms). Here is a snippet for future reference.

webpack.config.js
```
const CopyWebpackPlugin = require("webpack-plugin-copy");
const plugins = [
  new CopyWebpackPlugin([
    {
      from: "node_modules/apollo-engine-binary-linux",
      to: "node_modules/apollo-engine-binary-linux",
      copyPermissions: true
    },
    {
      from: "node_modules/apollo-engine",
      to: "node_modules/apollo-engine",
      copyPermissions: true
    }
  ])
];
const externals = [/^apollo-engine/];
```
